clis/teliod/src/daemon.rs

```diff
@@ -0,0 +1,242 @@
+use tracing::{debug, error, info, trace};
+
+use crate::core_api::{get_meshmap as get_meshmap_from_server, init_with_api};
+use crate::{
+    command_listener::{CommandListener, CommandResponse},
+    comms::DaemonSocket,
+    config::DeviceIdentity,
+    config::TeliodDaemonConfig,
+    error::TeliodError,
+    nc::NotificationCenter,
+    TelioStatusReport,
+};
+
+#[derive(Debug)]
+pub enum TelioTaskCmd {
+    // Command to set the downloaded meshmap to telio instance
+    UpdateMeshmap(MeshMap),
+    // Get telio status
+    GetStatus,
+    // Break the receive loop to quit the daemon and exit gracefully
+    Quit,
+}
+
+// From async context Telio needs to be run in separate task
+fn telio_task(
+    node_identity: Arc<DeviceIdentity>,
+    auth_token: Arc<String>,
+    mut rx_channel: mpsc::Receiver<TelioTaskCmd>,
+    tx_channel: mpsc::Sender<TelioTaskCmd>,
+    response_tx_channel: mpsc::Sender<CommandResponse>,
+    interface_name: &str,
+) -> Result<(), TeliodError> {
+    debug!("Initializing telio device");
+    let mut telio = Device::new(
+        Features::default(),
+        // TODO: replace this with some real event handling
+        move |event| info!("Incoming event: {:?}", event),
+        None,
+    )?;
+
+    // TODO: This is temporary to be removed later on when we have proper integration
+    // tests with core API. This is to not look for tokens in a test environment
+    // right now as the values are dummy and program will not run as it expects
+    // real tokens.
+    if !auth_token.as_str().eq("") {
+        start_telio(&mut telio, node_identity.private_key, interface_name)?;
+        task_retrieve_meshmap(node_identity, auth_token, tx_channel);
+
+        while let Some(cmd) = rx_channel.blocking_recv() {
+            info!("Got command {:?}", cmd);
+            match cmd {
+                TelioTaskCmd::UpdateMeshmap(map) => {
+                    if let Err(e) = telio.set_config(&Some(map)) {
+                        error!("Unable to set meshmap due to {e}");
+                    }
+                }
+                TelioTaskCmd::GetStatus => {
+                    let status_report = TelioStatusReport {
+                        is_running: telio.is_running(),
+                        nodes: telio.external_nodes()?,
+                    };
+                    debug!("Telio status: {:#?}", status_report);
+                    if let Err(e) = response_tx_channel
+                        .blocking_send(CommandResponse::StatusReport(status_report))
+                    {
+                        error!("Telio task failed sending response: {}", e)
+                    }
+                }
+                TelioTaskCmd::Quit => {
+                    telio.stop();
+                    break;
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn task_retrieve_meshmap(
+    device_identity: Arc<DeviceIdentity>,
+    auth_token: Arc<String>,
+    tx: mpsc::Sender<TelioTaskCmd>,
```

Comment:
feels to me like this should be a oneshot ?

Reply:
Good point! I didn't change this bit besides moving it, but I'll see if I can quickly convert it

Reply:
Hmmm or maybe not, if I understand correctly. Because the `telio_task` is running in a loop, the `TelioTaskCmd` `rx` part would be dropped after the first command was processed leaving the daemon unable to process further commands. Maybe @Hasan6979 can comment better as he designed this part?

What could possibly become a `oneshot` channel is the `CommandResponse` part, as every client command is expecting to have only one reply.  But in that case every task command would have to pass their own new `response_tx` part to the `telio_task`. Then the TelioTaskCmd would look something like this
```rust
TelioTaskCmd::GetStatus { response_tx: oneshot::Sender<TelioStatusReport> }
```
instead of 
```rust
TelioTaskCmd::GetStatus
```

My idea with `CommandResponse` channel was that it could be reused by all of the commands, for example if we want to return `Ok` status from `UpdateMeshmap`. What do you think?

Reply:
Or is it only the Sender that gets dropped after the first message in a `oneshot`?

Reply:
If I got it right

>  [oneshot channel](https://docs.rs/tokio/latest/tokio/sync/oneshot/index.html) supports sending a **single value** from a single producer to a single consumer. This channel is usually used to send the result of a computation to a waiter.

Reply:
Sender is consumed at `send` and it doesn't implement a `clone` trait as well. If `try_recv` is attempted again it would result in an error since the tx part is dropped.
We would want it to not be OneShot as it would still receive update commands from MQTT.